File: app/src/main/java/facebook/com/socialrunner/domain/service/RouteService.java
package facebook.com.socialrunner.domain.service;

import com.google.android.gms.maps.model.LatLng;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import facebook.com.socialrunner.domain.data.entity.Route;
import facebook.com.socialrunner.domain.data.entity.RoutePoint;
import facebook.com.socialrunner.domain.data.entity.Runner;
import facebook.com.socialrunner.domain.data.repository.ResultHandler;
import facebook.com.socialrunner.domain.data.repository.RouteRepository;
import facebook.com.socialrunner.domain.data.repository.RunnerRepository;
import facebook.com.socialrunner.util.backward.Consumer;

public class RouteService {

    private RouteRepository routeRepository;
    private RunnerRepository runnerRepository;

    // Dependency inversion compromised
    public RouteService() {
        this.routeRepository = new RouteRepository();
        this.runnerRepository = new RunnerRepository();
    }

    public void getRoutesForUser(String username, Consumer<List<Route>> onFetch) {

        ResultHandler<Runner> runnerHandler = new ResultHandler<>(runner -> {

            Map<String, Route> routes = new HashMap<>();

            for (String routeId : runner.getRouteIds()) {

                routeRepository.fetchById(routeId, new ResultHandler<Route>(route -> {
                    routes.put(routeId, route);
                    onFetch.accept(new ArrayList<>(routes.values()));
                }));
            }
        });

        runnerRepository.fetchByName(username, runnerHandler);
    }

    public void uploadNewRoute(String username, List<LatLng> waypoints) {

        ResultHandler<Runner> handler = new ResultHandler<>(runner -> {
            Route route = new Route();
            route.setRoutePoints(parseCoordinates(waypoints));

            routeRepository.create(route);

            if (runner != null) {
                runner.getRouteIds().add(route.getId());
                runnerRepository.update(runner);
            }
        });

        runnerRepository.fetchByName(username, handler);
    }

    private List<RoutePoint> parseCoordinates(List<LatLng> coordinates) {

        List<RoutePoint> routePoints = new ArrayList<>();

        for (LatLng coordinate : coordinates)
            routePoints.add(new RoutePoint(coordinate.latitude, coordinate.longitude));

        return routePoints;
    }
}

File: app/src/main/java/facebook/com/socialrunner/domain/service/RouteService.kt
package facebook.com.socialrunner.domain.service

import android.location.Location
import com.google.android.gms.maps.model.LatLng
import facebook.com.socialrunner.domain.data.entity.Position
import facebook.com.socialrunner.domain.data.entity.Route
import facebook.com.socialrunner.domain.data.entity.RoutePoint
import facebook.com.socialrunner.domain.data.entity.Runner
import facebook.com.socialrunner.domain.data.repository.ResultHandler
import facebook.com.socialrunner.domain.data.repository.RouteRepository
import facebook.com.socialrunner.domain.data.repository.RunnerRepository
import facebook.com.socialrunner.toRoutePoints
import facebook.com.socialrunner.toRoutes
import java.util.*


class RouteService {

    private val routeRepository: RouteRepository = RouteRepository()
    private val runnerRepository: RunnerRepository = RunnerRepository()

    fun getQueriesInArea(loc: LatLng, onFetch: (Route) -> Unit) {
        val loca = location(loc)
        val handler = ResultHandler<Map<String, Map<*, *>>> { routes ->
            routes?.run {
                toRoutes().filter { route -> route.routePoints.any { rp -> isInRange(rp, loca) } }
                        .forEach(onFetch)
            }
        }
        routeRepository.fetchFromRoutes(handler)
    }

    private fun isInRange(rp: RoutePoint, base: Location): Boolean {
        return location(rp).distanceTo(base) < 5000
    }

    private fun location(l: LatLng) = Location("").apply {
        latitude = l.latitude
        longitude = l.longitude
    }

    private fun location(rp: RoutePoint) = Location("").apply {
        latitude = rp.loc.latitude
        longitude = rp.loc.longitude
    }


    fun uploadNewRoute(username: String, route: Route, waypoints: MutableList<LatLng>) {
        route.routePoints.addAll(waypoints.toRoutePoints())

        val handler = ResultHandler<Runner> { runner ->
            //routeRepository.create(route)

//            if (runner != null) {
//                runner.routeIds.add(route.id!!)
//                runnerRepository.update(runner)
//            }
        }

        runnerRepository.fetchByName(username, handler)
    }

    private fun parseCoordinates(coordinates: MutableList<LatLng>) =
            coordinates.map { RoutePoint(loc = Position(it.latitude, it.longitude)) }
                    .toMutableList()
}
